app/src/main/java/ee/ioc/phon/android/speechutils/AbstractAudioRecorder.java

```diff
@@ -128,20 +152,71 @@ protected int getStatus(int numOfBytes, int len) {
     }
 
     /**
+     * Check if the consume pointer was crossed by the recorded pointer. As long as the consume
+     * pointer was not crossed, the consumption of the buffer may continue as usual and no sound gap
+     * will occur. Once the consume pointer was crossed (e.g. it was on sample 1000 and prior to this
+     * read the recorder was on sample 750 and now that it read the new sample it's on 1500), there's
+     * an audio gap between the consumer and the recorder that can not be filled (data is lost with
+     * no ability to get it back). Whenever this kind of cross occurs, the calling code changed the
+     * session id of the recorder so that if consume is called (from ContinuousRawAudioRecorder),
+     * it will not assume that the data is complete and could be fetched but it will act according to
+     * the SessionStartPointer configured (e.g. read the buffer from the beginning, from now, or from
+     * now - X millis)
+     * @param reachedTheEndOfRecordingBuffer - in case that in the read before the call to this method the recorder
+     *                       passed the end of the buffer and returned to the beginning
+     * @param numOfBytesRead - in the reading process
+     * @return true/false according to the above logic
+     */
+    private boolean isConsumePointerCrossed(boolean reachedTheEndOfRecordingBuffer, int numOfBytesRead) {
+        return numOfBytesRead > 0 &&
+               mRecordingBufferIsFullWithData &&
+               isRecordedSessionSameAsConsumedSession() &&
+               ((mRecordedLength - numOfBytesRead < mConsumedLength && mRecordedLength >= mConsumedLength) ||
+                (reachedTheEndOfRecordingBuffer && mConsumedLength < mRecordedLength));
+    }
+
+    /**
      * Copy data from the given recorder into the given buffer, and append to the complete recording.
      * public int read (byte[] audioData, int offsetInBytes, int sizeInBytes)
      */
     protected int read(SpeechRecord recorder, byte[] buffer) {
         int len = buffer.length;
         int numOfBytes = recorder.read(buffer, 0, len);
         int status = getStatus(numOfBytes, len);
-        if (status == 0) {
-            // arraycopy(Object src, int srcPos, Object dest, int destPos, int length)
-            // numOfBytes <= len, typically == len, but at the end of the recording can be < len.
-            System.arraycopy(buffer, 0, mRecording, mRecordedLength, numOfBytes);
-            mRecordedLength += len;
+        boolean reachedTheEndOfRecordingBuffer = false;
+        // if we need to keep on listening, when reaching the end of the recorded buffer,
+        // continue to write from the beginning. thus, we have a cyclic buffer
+        if (mAlwaysListen && status == -300) {
+            reachedTheEndOfRecordingBuffer = true;
+            status = 0;
```

Comment:
in general, in continuous mode, we don't want to stop listening. we want to fix the situation and continue